src/io.c

```diff
@@ -140,67 +140,45 @@ unsigned io_fprintln(const char *fname, const char *fmt, ...)
  * is created.
  * The datadir argument can be used to specify an alternative data root dir.
  * The confdir argument can be used to specify an alternative configuration dir.
+ *
+ * The statically allocated buffers (in vars.c) have BUFSIZ bytes for each path.
+ * Allow BUFSIZ - 20 for the externally supplied directory part.
  */
 void io_init(const char *cfile, const char *datadir, const char *confdir)
 {
-	const char *home;
-	const char *conf_home;
-
-	if (datadir != NULL) {
-		home = datadir;
-		if (confdir == NULL)
-			conf_home = home;
-		else
-			conf_home = confdir;
-
-		snprintf(path_dir, BUFSIZ, "%s", home);
-
-		snprintf(path_conf_dir, BUFSIZ, "%s",  conf_home);
-		snprintf(path_conf, BUFSIZ, "%s/" CONF_PATH_NAME, conf_home);
-		snprintf(path_keys, BUFSIZ, "%s/" KEYS_PATH_NAME, conf_home);
-		snprintf(path_hooks, BUFSIZ, "%s/" HOOKS_DIR_NAME, conf_home);
-
-		snprintf(path_todo, BUFSIZ, "%s/" TODO_PATH_NAME, home);
-		snprintf(path_cpid, BUFSIZ, "%s/" CPID_PATH_NAME, home);
-		snprintf(path_dpid, BUFSIZ, "%s/" DPID_PATH_NAME, home);
-		snprintf(path_notes, BUFSIZ, "%s/" NOTES_DIR_NAME, home);
-		snprintf(path_dmon_log, BUFSIZ, "%s/" DLOG_PATH_NAME, home);
-	} else {
-		home = getenv("HOME");
-		if (home == NULL) {
-			home = ".";
-		}
-		if (confdir == NULL)
-			conf_home = home;
-		else
-			conf_home = confdir;
-
-		snprintf(path_dir, BUFSIZ, "%s/" DIR_NAME, home);
-
-		snprintf(path_conf_dir, BUFSIZ, "%s",  conf_home);
-		snprintf(path_conf, BUFSIZ, "%s/" CONF_PATH, conf_home);
-		snprintf(path_keys, BUFSIZ, "%s/" KEYS_PATH, conf_home);
-		snprintf(path_hooks, BUFSIZ, "%s/" HOOKS_DIR, conf_home);
-
-		snprintf(path_todo, BUFSIZ, "%s/" TODO_PATH, home);
-		snprintf(path_cpid, BUFSIZ, "%s/" CPID_PATH, home);
-		snprintf(path_dpid, BUFSIZ, "%s/" DPID_PATH, home);
-		snprintf(path_notes, BUFSIZ, "%s/" NOTES_DIR, home);
-		snprintf(path_dmon_log, BUFSIZ, "%s/" DLOG_PATH, home);
-	}
+#define SNPRINTF(d, s1, s2) \
+  EXIT_IF(snprintf(d, BUFSIZ - 20, "%s%s", s1, s2) >= BUFSIZ - 20, \
+          _("Directory name too long: %s (limit %d)"), s1, BUFSIZ - 21);
```

Comment:
This is a very non-performance critical data path. Can we simply use asprintf() to avoid these magic numbers and potential pitfalls with buffer sizes?

Reply:
Done in new commit.